storage-proofs/src/merkle/proof.rs

```diff
@@ -0,0 +1,995 @@
+    fn try_from_proof(p: proof::Proof<<H as Hasher>::Domain, Arity>) -> Result<Self> {
+        Ok(proof_to_single(&p, 1, None))
+    }
+
+    fn verify(&self) -> bool {
+        let calculated_root = self.path.root(self.leaf);
+        self.root == calculated_root
+    }
+
+    fn leaf(&self) -> H::Domain {
+        self.leaf
+    }
+
+    fn root(&self) -> H::Domain {
+        self.root
+    }
+
+    fn len(&self) -> usize {
+        self.path.len() * (Arity::to_usize() - 1) + 2
+    }
+
+    fn path(&self) -> Vec<(Vec<H::Domain>, usize)> {
+        self.path
+            .iter()
+            .map(|x| (x.hashes.clone(), x.index))
+            .collect::<Vec<_>>()
+    }
+
+    fn path_index(&self) -> usize {
+        self.path.path_index()
+    }
+
+    fn proves_challenge(&self, challenge: usize) -> bool {
+        self.path_index() == challenge
+    }
+}
+
+impl<H: Hasher, Arity: 'static + PoseidonArity, SubTreeArity: 'static + PoseidonArity>
+    SubProof<H, Arity, SubTreeArity>
+{
+    fn try_from_proof(p: proof::Proof<<H as Hasher>::Domain, Arity>) -> Result<Self> {
+        ensure!(
+            p.sub_layer_nodes() == SubTreeArity::to_usize(),
+            "sub arity mismatch"
+        );
+        ensure!(
+            p.sub_tree_proof.is_some(),
+            "Cannot generate sub proof without a base-proof"
+        );
+        let base_p = p.sub_tree_proof.as_ref().unwrap();
+
+        // Generate SubProof
+        let root = p.root();
+        let leaf = base_p.item();
+        let base_proof = extract_path::<H, Arity>(base_p.lemma(), base_p.path(), 1);
+        let sub_proof = extract_path::<H, SubTreeArity>(p.lemma(), p.path(), 0);
+
+        Ok(SubProof::new(base_proof, sub_proof, root, leaf))
+    }
+
+    fn verify(&self) -> bool {
+        let sub_leaf = self.base_proof.root(self.leaf);
+        let calculated_root = self.sub_proof.root(sub_leaf);
+
+        self.root == calculated_root
+    }
+
+    fn leaf(&self) -> H::Domain {
+        self.leaf
+    }
+
+    fn root(&self) -> H::Domain {
+        self.root
+    }
+
+    fn len(&self) -> usize {
+        SubTreeArity::to_usize()
+    }
+
+    fn path(&self) -> Vec<(Vec<H::Domain>, usize)> {
+        self.base_proof
+            .iter()
+            .map(|x| (x.hashes.clone(), x.index))
+            .chain(self.sub_proof.iter().map(|x| (x.hashes.clone(), x.index)))
+            .collect()
+    }
+
+    fn path_index(&self) -> usize {
+        let mut base_proof_leaves = 1;
+        for _i in 0..self.base_proof.len() {
+            base_proof_leaves *= Arity::to_usize()
+        }
+
+        let sub_proof_index = self.sub_proof.path_index();
+
+        (sub_proof_index * base_proof_leaves) + self.base_proof.path_index()
+    }
+
+    fn proves_challenge(&self, challenge: usize) -> bool {
+        let sub_path_index =
+            ((challenge / Arity::to_usize()) * Arity::to_usize()) + (challenge % Arity::to_usize());
+
+        sub_path_index == challenge
```

Comment:
Blocking this always returns true, making the entire function misleading — and any usage almost certainly wrong.

To see that the result calculated here cannot be correct, consider that `SubTreeArity` is unused in the computation.